### PR TITLE
POC GatewayAPI HTTPRoute timeout

### PR DESCRIPTION
**Pull request by SRodi**
**Please provide a description of this PR:**

Implementation to support POC on HTTRoute Timeouts for GatewayAPI 

references:
https://github.com/kubernetes-sigs/gateway-api/pull/2013
https://gateway-api.sigs.k8s.io/geps/gep-1742/
https://gateway-api.sigs.k8s.io/geps/gep-2257/
https://github.com/kubernetes-sigs/gateway-api/issues/1742
https://github.com/kubernetes-sigs/gateway-api/pull/1997

### CODE REVIEW

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 1]
```diff
@@ -291,6 +292,12 @@ func buildHTTPVirtualServices(
 			vs.Route = route
 		}
 
+		tr, err := createTimeoutRequest(*r.Timeouts)
+		if err != nil {
+			reportError(err)
+		}
+		vs.Timeout = tr.Timeout
```
Comment (frankbu) [inline pilot/pkg/config/kube/gateway/conversion.go:299]: I think this should just be something like this?
```suggestion
		if r.Timeouts != nil && r.Timeouts.Request != nil {
			vs.Timeout = durationpb.New(r.Timeouts.Request.Duration)
		}
```

Comment (SRodi) [reply pilot/pkg/config/kube/gateway/conversion.go:299]: @frankbu do we still need `createTimeoutRequest`  function to separate in future from `workloadRequest` timeout as per https://github.com/kubernetes-sigs/gateway-api/pull/1997#discussion_r1189088248 ?

Comment (frankbu) [reply pilot/pkg/config/kube/gateway/conversion.go:299]: No. I think you're confusing istio HTTPRoute with gateway-api HTTPRoute. We've already split timeout out of gateway-api HTTPRoute as suggested in the comment (previously I had `spec.rules.timeout`, but now it is `spec.rules.timeouts.request`). If you look at the latest [GEP-1742 PR](https://github.com/kubernetes-sigs/gateway-api/pull/1997) you'll see that I've also added `backendRequest` (renamed from `workloadRequest`) under `spec.rules.timeouts`.

Comment (frankbu) [reply pilot/pkg/config/kube/gateway/conversion.go]: If you would like to pencil in the code for `backendRequest`, I think the above code would be something like this:
```
		if r.Timeouts != nil {
			if r.Timeouts.Request != nil {
				vs.Timeout = durationpb.New(r.Timeouts.Request.Duration)
			}
			// TODO
			// if r.Timeouts.BackendRequest != nil {
				// if vs.Retries == nil {
				//	vs.Retries = create new one
				// }
				// vs.Retries.PerTryTimeout = durationpb.New(r.Timeouts.BackendRequest.Duration)
			// }
		}
```

**Comment by frankbu** [comment]
@SRodi Please add a unit test to confirm that a route with timeout produces a timeout in the generated VirtualService.

https://github.com/istio/istio/blob/master/pilot/pkg/config/kube/gateway/conversion_test.go
https://github.com/istio/istio/blob/master/pilot/pkg/config/kube/gateway/testdata/http.yaml

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 2]
```diff
@@ -291,6 +292,10 @@ func buildHTTPVirtualServices(
 			vs.Route = route
 		}
 
+		if r.Timeouts != nil && r.Timeouts.Request != nil {
+			vs.Timeout = durationpb.New(r.Timeouts.Request.Duration)
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/gateway/conversion.go:296]: When should we use `BackendRequest`? Looks it is outbound time, which should be set to vs timeout

Comment (frankbu) [reply pilot/pkg/config/kube/gateway/conversion.go:296]: `BackendRequest` maps to VS `rules.retires.perTryTimeout`. Here is the draft GEP this is implementing: https://github.com/kubernetes-sigs/gateway-api/blob/ad0e9c7a569bde169254a47a68e7f760ad6ebb15/geps/gep-1742.md#api

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:296]: so we can support BackendRequest, right? we just don't in this PR, but we can

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 3]
```diff
@@ -292,6 +293,15 @@ func buildHTTPVirtualServices(
 			vs.Route = route
 		}
 
+		if r.Timeouts != nil {
+			if r.Timeouts.Request != nil {
+				vs.Timeout = durationpb.New(r.Timeouts.Request.Duration)
+			}
+			if r.Timeouts.BackendRequest != nil {
+				vs.Retries.PerTryTimeout = durationpb.New(r.Timeouts.BackendRequest.Duration)
```
Comment (frankbu) [inline pilot/pkg/config/kube/gateway/conversion.go:301]:
```suggestion
				if vs.Retries == nil {
					vs.Retries = &istio.HTTPRetry{}
				}
				vs.Retries.PerTryTimeout = durationpb.New(r.Timeouts.BackendRequest.Duration)
```

Comment (SRodi) [reply pilot/pkg/config/kube/gateway/conversion.go:304]: addressed in 14443afd7b4a012c35984c3e8242daca517528e7

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 4]
```diff
@@ -292,6 +293,18 @@ func buildHTTPVirtualServices(
 			vs.Route = route
 		}
 
+		if r.Timeouts != nil {
+			if r.Timeouts.Request != nil {
+				vs.Timeout = durationpb.New(r.Timeouts.Request.Duration)
+			}
+			if r.Timeouts.BackendRequest != nil {
+				if vs.Retries == nil {
+					vs.Retries = &istio.HTTPRetry{Attempts: 1}
```
Comment (frankbu) [inline pilot/pkg/config/kube/gateway/conversion.go:302]: This is not quite right. Setting a BackendRequest timeout will have an undesirable side effect of doing 1 retry. The problem is that Istio complains if the PerTryTimeout is set when Attempts is 0: `configuration is invalid: http retry policy configured when attempts are set to 0 (disabled)`.

@howardjohn Do you think we can change the validation to allow PerTryTimeout with Attempts=0, or will Envoy also not support that?

I suppose we could handle this by setting vs.Timeout to the value of BackendTimeout in the case that vs.Retries == nil?

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:302]: > I suppose we could handle this by setting vs.Timeout to the value of BackendTimeout in the case that vs.Retries == nil?

:+1: 

BTW, do we validate BackendTimeout >= PerTryTimeout?

Comment (frankbu) [reply pilot/pkg/config/kube/gateway/conversion.go:302]: > BTW, do we validate BackendTimeout >= PerTryTimeout?

Not sure what you are asking?

The gateway-api validator does check that timeouts.BackendRequest <= timeouts.Request, which means that vs.Retries.PerTryTimout will always be <= vs.Timeout.

Comment (frankbu) [reply pilot/pkg/config/kube/gateway/conversion.go:302]: Oh, I see now, that is what you were asking, so the answer is yes 😃

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 5]
```diff
@@ -292,6 +293,18 @@ func buildHTTPVirtualServices(
...
+				}
+				vs.Retries.PerTryTimeout = durationpb.New(r.Timeouts.BackendRequest.Duration)
```
Comment (frankbu) [inline pilot/pkg/config/kube/gateway/conversion.go:304]:
```suggestion
				timeout := durationpb.New(r.Timeouts.BackendRequest.Duration)
				if vs.Retries != nil {
					vs.Retries.PerTryTimeout = timeout
				} else {
					vs.Timeout = timemout
				}
```

Comment (SRodi) [reply pilot/pkg/config/kube/gateway/conversion.go]: addressed in a6c6566e5675613b5ea7215f0c4d9546a952aa34

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 6]
```diff
@@ -292,6 +293,20 @@ func buildHTTPVirtualServices(
 			vs.Route = route
 		}
 
+		if r.Timeouts != nil {
+			if r.Timeouts.Request != nil {
```
Comment (frankbu) [inline pilot/pkg/config/kube/gateway/conversion.go:297]: Need to update this now that GEP-1742 timeout value 0 -> disable.
```suggestion
			if r.Timeouts.Request != nil && r.Timeouts.Request.Duration != 0 {
```

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 7]
```diff
@@ -292,6 +293,20 @@ func buildHTTPVirtualServices(
...
+				vs.Timeout = durationpb.New(r.Timeouts.Request.Duration)
+			}
+			if r.Timeouts.BackendRequest != nil {
```
Comment (frankbu) [inline pilot/pkg/config/kube/gateway/conversion.go:300]:
```suggestion
			if r.Timeouts.BackendRequest != nil && r.Timeouts.BackendRequest.Duration != 0 {
```

**Comment by frankbu** [comment]
not stale

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 8]
```diff
@@ -247,6 +248,21 @@ func convertHTTPRoute(r k8s.HTTPRouteRule, ctx configContext,
...
+		}
+		if r.Timeouts.BackendRequest != nil && r.Timeouts.BackendRequest.Duration != 0 {
+			timeout := durationpb.New(r.Timeouts.BackendRequest.Duration)
+			if vs.Retries != nil {
+				vs.Retries.PerTryTimeout = timeout
+			} else {
+				vs.Timeout = timeout
+			}
+		}
+	}
```
Comment (frankbu) [inline pilot/pkg/config/kube/gateway/conversion.go:264]:
```suggestion
	if r.Timeouts != nil {
		if r.Timeouts.Request != nil {
			request, _ := time.ParseDuration((string)(*r.timeouts.Request))
			if request != 0 {
				vs.Timeout = durationpb.New(request)
			}
		}
		if r.Timeouts.BackendRequest != nil {
			backendRequest, _ := time.ParseDuration((string)(*r.timeouts.BackendRequest))
			if backendRequest != 0 {
				timeout := durationpb.New(backendRequest)
				if vs.Retries != nil {
					vs.Retries.PerTryTimeout = timeout
				} else {
					vs.Timeout = timeout
				}
			}
		}
	}
```

File: go.mod [hunk 9]
```diff
@@ -250,3 +250,5 @@ require (
 	sigs.k8s.io/kustomize/kyaml v0.14.3 // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.3.0 // indirect
 )
+
+replace sigs.k8s.io/gateway-api => /Users/simone/go/src/sigs.k8s.io/gateway-api
```
Comment (SRodi) [inline go.mod:254]: I will remove the `replace` once the GatewayAPI PR https://github.com/kubernetes-sigs/gateway-api/pull/2013 will be merged and Timeouts implementation will be part of `k8sbeta "sigs.k8s.io/gateway-api/apis/v1beta1"`

**Comment by SRodi** [comment]
this can be closed - see https://github.com/istio/istio/pull/47017

**Pull request closed without merging**